lib/Sema/CSDiagnostics.cpp

```diff
@@ -3528,6 +3527,66 @@ ContextualFailure::getDiagnosticFor(ContextualTypePurpose context,
   return None;
 }
 
+bool NonClassTypeToAnyObjectConversionFailure::diagnoseAsError() {
+  if (RawLocator->isForContextualType()) {
+    ContextualFailure failure(getSolution(), RawFromType, RawToType,
+                              RawLocator);
+    return failure.diagnoseAsError();
+  }
+
+  if (RawLocator->isLastElement<LocatorPathElt::ApplyArgToParam>()) {
+    ArgumentMismatchFailure failure(getSolution(), RawFromType, RawToType,
+                                    RawLocator);
+    return failure.diagnoseAsError();
+  }
+
+  Optional<Diag<Type, Type>> diagnosis;
+  bool forProtocol = RawToType->isConstraintType();
+  auto originalAnchor = getLocator()->getAnchor();
+
+  if (isExpr<ArrayExpr>(originalAnchor)) {
+    diagnosis = forProtocol ? diag::cannot_convert_array_element_protocol
+                            : diag::cannot_convert_array_element;
+  } else if (isExpr<DictionaryExpr>(originalAnchor)) {
+    auto lastElem =
+        RawLocator->getPath().back().castTo<LocatorPathElt::TupleElement>();
```

Comment:
Nit: Use `getLastElementAs<LocatorPathElt::TupleElement>`. We also have `isLastElement` and `castLastElementTo` .